xds: Allow endpoints to wait for the current policy version to be acked

jrajahalme opened the pull request:
Even when no policy change is needed, the endpoint may need to wait
for the current policy version to be acked by the proxy.

Use a different Envoy node ID for proxylib, so that proxylib ACK
arriving before the main Envoy ACK does not trigger DNS response to be
released to the source POD.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9395)
<!-- Reviewable:end -->

aanm reviewed (commented):
@jrajahalme it seems the tests have a legit failure

aanm commented on pkg/envoy/xds/ack.go:
```diff
@@ -163,7 +169,20 @@ func (m *AckingResourceMutatorWrapper) Upsert(typeURL string, resourceName strin
+	m.locker.Lock()
+	defer m.locker.Unlock()
+
+	if wg != nil {
```
It seems weird to have this check here, why do we have it?

jrajahalme replied:
It was there to avoid using a nil pointer, but for this code path the pointer is never nil, so I removed the check and added comments to the callers that the WaitGroup may not be nil.

On the corresponding UpdateNetworkPolicy path we explicitly set `wg` to `nil` when we know that there is no proxy running (yet) to ACK the changes.

ianvernon commented on pkg/envoy/server.go:
```diff
@@ -787,36 +787,18 @@ func (s *XDSServer) UpdateNetworkPolicy(ep logger.EndpointUpdater, policy *polic
+	nodeIDs := make([]string, 0, 1)
+	if ep.HasSidecarProxy() {
+		if ep.GetIPv4Address() == "" {
+			log.Fatal("Envoy: Sidecar proxy has no IPv4 address")
```
Do we not support IPv6-only mode when running alongside Istio?

jrajahalme replied:
So far Istio support has been IPv4 only, actually don't know if/when k8s and Istio support IPv6-only mode.

ianvernon commented on pkg/envoy/resources.go:
```diff
@@ -60,7 +59,7 @@ var (
 
 // HandleResourceVersionAck is required to implement ResourceVersionAckObserver.
-func (cache *NPHDSCache) HandleResourceVersionAck(ackVersion uint64, nackVersion uint64, node *envoy_api_v2_core.Node, resourceNames []string, typeURL string, detail string) {
+func (cache *NPHDSCache) HandleResourceVersionAck(ackVersion uint64, nackVersion uint64, nodeIP string, resourceNames []string, typeURL string, detail string) {
```
Add a comment about what an empty `nodeIP` string signifies since callers can pass an empty string here.

jrajahalme replied:
Callers of this never pass an empty `nodeIP`. I added a comment to explain what the function does, though.

ianvernon commented on pkg/envoy/xds/cache.go:
```diff
@@ -212,14 +222,27 @@ func (c *Cache) Clear(typeURL string, force bool) (version uint64, updated bool)
-func (c *Cache) GetResources(ctx context.Context, typeURL string, lastVersion uint64,
-	node *envoy_api_v2_core.Node, resourceNames []string) (*VersionedResources, error) {
+// Ack updates the last seen ACKed version for the node
+func (c *Cache) Ack(ackedVersion uint64, nodeIP string) {
```
Add unit tests for this function.

jrajahalme replied:
Added unit tests

ianvernon commented on pkg/envoy/xds/server.go:
```diff
@@ -309,13 +313,19 @@ func (s *Server) processRequestStream(ctx context.Context, streamLog *logrus.Ent
 			state := &typeStates[index]
+			watcher := s.watchers[typeURL]
+
+			nodeIP, err := IstioNodeToIP(req.GetNode())
```
Should this `IstioNodeToIP` function be renamed?

jrajahalme replied:
Not really, it is the Istio convention for structuring the Envoy Node ID that we use to be able to derive the IP address of the node. We control the node ID format for the host proxy ourselves, so using the Istio format also for the host proxy allows more uniform handling of the node ID.

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -438,6 +438,16 @@ func (l4 *L4Filter) IsRedirect() bool {
 	return l4.L7Parser != ParserTypeNone
 }
 
+// IsEnvoyRedirect returns true if the L4 filter contains a port redirected to Envoy
```
Add unit tests for the functions added in this file.

jrajahalme replied:
Added to existing functions

jrajahalme committed:
logfields: Add tag for cached xDS version.

Add and use a tag for cached xDS version.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
Envoy: Track last ACKed version per proxy node

Track ACKs sent by proxy nodes. This is needed to be able to wait for
the current cached resource version to be realized by all relevant
nodes.

Node entry in the map of acked versions is deleted when the network
policy for the endpoint is removed, which happens when the endpoint is
deleted.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Always use IstioNodeToIP function

Envoy Node ID to IP mapping function is configurable, but we always
use the Istio mapping anyway. Rather than spread the configurability
elsewhere, always use the Istio convention.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
xds: Allow endpoints to wait for the current policy version to be acked

Even when no policy change is needed, the endpoint may need to wait
for the current policy version to be acked by the proxy.

Use a different Envoy node ID for proxylib, so that proxylib ACK
arriving before the main Envoy ACK does not trigger DNS response to be
released to the source POD.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Do not force Network Policy updates

Use the new "UseCurrent" functionality to not force newtwork policy
updates to be sent to Envoy when the policy did not change. Previously
the 'force' flag was necessary, as there was no other way to wait for
the ACK of the current version.

This saves some unnecessary signaling between cilium-agent and Envoy.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Add unit tests

Add tests of the new l4 functions to the existing unit tests.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Remove 'force' argument from cache operations

Now that 'force' argument is never set to 'true' it can be removed.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

ianvernon commented on pkg/envoy/server.go:
```diff
@@ -786,36 +786,18 @@ func (s *XDSServer) UpdateNetworkPolicy(ep logger.EndpointUpdater, policy *polic
-		// TODO: Remove this when we implement and inject an Envoy network filter
-		// into every Envoy listener to filter at L3/L4.
-		var hasL7Rules bool
-	Policies:
```
Can you elaborate on why it is safe to remove this check whose purpose is to set the waitgroup to nil if we are running with a sidecar proxy?

aanm commented on proxylib/npds/client_test.go:
```diff
@@ -56,17 +57,21 @@ var resources = []*cilium.NetworkPolicy{
 // UpsertNetworkPolicy must only be used for testing!
 func (cs *ClientSuite) UpsertNetworkPolicy(c *C, s *envoy.XDSServer, p *cilium.NetworkPolicy) {
-	comp := completion.NewCompletion(nil, func(err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
Why 5 seconds and not 10 for example?

jrajahalme replied:
No reason, most likely I copied this from somewhere else in the test. 5 seconds should be long enough given that both client and server run locally within the test.
